File: modules/tools/npm.go
package tools

import (
	"os"
	"os/exec"
	"path"
	"strings"
)

// ---------------------------------
// Vars

// ---------------------------------
// Public functions

// NpmInstall installs node dependencies
func NpmInstall(deps []string, cmdDir string) (log string, err error) {
	vendorPath := NpmFindModules(cmdDir)

	for _, dep := range deps {
		realDep := NpmGetDepName(dep)

		// Install
		if NotExist(path.Join(vendorPath, realDep)) {
			// Create an *exec.Cmd
			cmd := exec.Command("npm", "install", dep)
			cmd.Stdout = os.Stdout
			cmd.Stderr = os.Stderr

			// Execute command
			cmd.Run()
		}
	}

	return
}

// NpmFindModules tries to find a node_modules folder
func NpmFindModules(cmdDir string) (vendorPath string) {
	basePath := cmdDir
	dirFound := false

	// Lets try and find related to the CmdDir
	for i := 0; i < 5; i++ {
		vendorPath = path.Join(basePath, "node_modules")

		if NotExist(vendorPath) {
			basePath = path.Join(basePath, "..")
		} else {
			dirFound = true
			break
		}
	}

	// Lets try and find related to the working dir
	if !dirFound {
		basePath, _ = os.Getwd()

		for i := 0; i < 5; i++ {
			vendorPath = path.Join(basePath, "node_modules")

			if NotExist(vendorPath) {
				basePath = path.Join(basePath, "..")
			} else {
				dirFound = true
				break
			}
		}
	}

	return vendorPath
}

// NpmGetDepName returns the dependency name
func NpmGetDepName(dep string) (name string) {
	i := strings.Index(dep, "@")
	if i == -1 {
		i = strings.Index(dep, "~")
	}
	if i == -1 {
		i = strings.Index(dep, "^")
	}
	if i == -1 {
		i = len(dep)
	}

	// Now the name
	name = dep[0:i]

	return name
}

File: modules/tools/npm.js
'use strict';

// -----------------------------------------
// IMPORTS

var path = require('path');
var exec = require('sync-exec');
var Joi = require('joi');
var validate = require('./validate.js');
var log = require('./log');
var general = require('./general');

// -----------------------------------------
// VARS

// -----------------------------------------
// PUBLIC FUNCTIONS

/**
 * Tries to find a node_modules folder
 * @return {string}
 */
function npmFindModules() {
    var basePath = process.cwd();
    var vendorPath;
    var i;

    // Lets try and find the node_modules
    for (i = 0; i < 5; i += 1) {
        vendorPath = path.join(basePath, 'node_modules');

        if (general.notExist(vendorPath)) {
            basePath = path.join(basePath, '..');
        } else {
            // Found the dir!
            break;
        }
    }

    return vendorPath;
}

/**
 * Get the dependency name
 * @param  {string} dep
 * @return {string}
 */
function npmGetDepName(dep) {
    var i;

    validate.type(
        { dep: dep },
        { dep: Joi.string() }
    );

    i = dep.indexOf('@');
    if (i === -1) {
        i = dep.indexOf('~');
    }
    if (i === -1) {
        i = dep.indexOf('^');
    }
    if (i === -1) {
        i = dep.indexOf('#');
    }
    if (i === -1) {
        i = dep.length;
    }

    return dep.slice(0, i);
}

/**
 * Installs node dependencies
 * @param  {array} deps
 */
function npmInstall(deps) {
    var vendorPath;

    validate.type(
        { deps: deps },
        { deps: Joi.array().items(Joi.string()) }
    );

    vendorPath = npmFindModules();
    deps.forEach(function (dep) {
        var realDep = npmGetDepName(dep);
        var data;

        if (!general.notExist(path.join(vendorPath, realDep))) {
            return;
        }

        // Install
        data = exec('npm install ' + dep);

        if (data.stderr && data.stderr.length) {
            log.logErr(data.stderr);
        } else if (data.stdout && data.stdout.length) {
            log.log(data.stdout);
        }
    });
}

// -----------------------------------------
// EXPORTS

module.exports = {
    npmInstall: npmInstall,
    npmFindModules: npmFindModules,
    npmGetDepName: npmGetDepName
};
